### ISSUE TITLE
rp2: Occasional UART failures

### ISSUE THREAD

**Issue by peterhinch** [micropython/micropython#8344]
On RP2 the following script fails eventually, after between 3K and 13K messages. It runs indefinitely on a Pyboard 1.1. Traceback:
```python
msg0 sent 13370 got 13370     msg1 sent 12741 got 12741
msg0 sent 13380 got 13380     msg1 sent 12750 got 12749
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rats13.py", line 38, in <module>
  File "uasyncio/core.py", line 1, in run
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in run_until_complete
  File "rats13.py", line 36, in main
  File "rats13.py", line 23, in receiver
  File "uasyncio/stream.py", line 1, in readline
TypeError: unsupported types for __iadd__: 'bytes', 'NoneType'
>>> 
```
I also tested at 38400 baud with the same outcome. In some hours of total running I never saw a corrupted or missing message, but this failure was consistent. I guess the error is [here in stream.py](https://github.com/micropython/micropython/blob/5f50f4a130e60c4bc5f4fbf58ac849dcaf764e72/extmod/uasyncio/stream.py#L54).

Script - see comment for failing line:
```python
import uasyncio as asyncio
from machine import UART
from sys import exit

uart = UART(0, 9600)

msg = (b'sender() bbbbbbbbbbbbbb\n', b'other() aaaaaa\n')
cnt = [0, 0, 0, 0, 0]

async def sender(n, t):
    swriter = asyncio.StreamWriter(uart, {})
    while True:
        swriter.write(msg[n])
        await swriter.drain()
        cnt[n + 3] += 1
        await asyncio.sleep(t)

async def receiver():
    sreader = asyncio.StreamReader(uart)
    while True:
        res = await sreader.readline()  # TypeError: unsupported types for __iadd__: 'bytes', 'NoneType'
        try:
            idx = msg.index(res)
        except ValueError:
            print('Bad string received', res)
            break  # This never happens
        cnt[idx] += 1
        if not cnt[0] % 10:
            print(f"msg0 sent {cnt[3]} got {cnt[0]}     msg1 sent {cnt[4]} got {cnt[1]}")

async def main():
    asyncio.create_task(sender(0, 0.1))
    asyncio.create_task(sender(1, 0.105))
    await receiver()

asyncio.run(main())
```

**Comment by peterhinch** [comment]
See https://github.com/micropython/micropython/pull/8373 for an explanation and fix.

**Comment by dpgeorge** [comment]
Running this test, have I hit the "break" that is commented as never happening, I get:
```
...
msg0 sent 20940 got 20940     msg1 sent 19958 got 19958
msg0 sent 20940 got 20940     msg1 sent 19959 got 19959
Bad string received b'sender()b bbbbbbbbbbbbbb\n'
>>> 
```

**Comment by peterhinch** [comment]
Could this be another manifestation of the bug?

**Comment by dpgeorge** [comment]
Yes it could be.  It's definitely a bug!

**Comment by dpgeorge** [comment]
Fixed by d242a9b7f7b6d4e744889d5af6c17dfb54aed8c3
